include/oneapi/dpl/pstl/hetero/dpcpp/execution_sycl_defs.h

```diff
@@ -33,6 +37,92 @@ namespace execution
 inline namespace __dpl
 {
 
+namespace __internal
+{
+
+#if _ONEDPL_PREDEFINED_POLICIES
+struct __global_instance_tag {};
+#endif
+using __queue_factory = sycl::queue (*)();
+
+class alignas(sycl::queue) __queue_holder
+{
+    static_assert(sizeof(sycl::queue) >= sizeof(std::pair<std::uintptr_t, __queue_factory>));
+    static_assert(alignof(sycl::queue) >= alignof(std::uintptr_t));
+
+    union
+    {
+        sycl::queue __q;
+        std::pair<std::uintptr_t, __queue_factory> __flag_and_factory;
+    };
+
+    bool
+    __has_queue() const
+    {
+        bool res = true;
+#if _ONEDPL_PREDEFINED_POLICIES
+        // If the first size-of-pointer bytes are zeros, we consider there is no valid queue.
+        res = (__flag_and_factory.first != 0);
+        std::atomic_signal_fence(std::memory_order_acq_rel); // to prevent possible reordering due to type punning
+#endif
+        return res;
+    }
+
+  public:
+    template <typename T, typename = std::enable_if_t<std::is_constructible_v<sycl::queue, T&&>>>
```

Comment:
As for me, it's a little bit strange that we change params pack to one param here but checking is constructor available.
From one point of view, we know how we going to use this code: only with one param and this check isn't really required. From another point of view, if we think we should know the usage of this code, params pack looks better: for example, because `sycl::queue` has a lot of formats of constructor:
```C++
  explicit queue(const property_list& propList = {});

  explicit queue(const async_handler& asyncHandler,
                 const property_list& propList = {});

  template <typename DeviceSelector>
  explicit queue(const DeviceSelector& deviceSelector,
                 const property_list& propList = {});

  template <typename DeviceSelector>
  explicit queue(const DeviceSelector& deviceSelector,
                 const async_handler& asyncHandler,
                 const property_list& propList = {});

  explicit queue(const device& syclDevice, const property_list& propList = {});

  explicit queue(const device& syclDevice, const async_handler& asyncHandler,
                 const property_list& propList = {});

  template <typename DeviceSelector>
  explicit queue(const context& syclContext,
                 const DeviceSelector& deviceSelector,
                 const property_list& propList = {});

  template <typename DeviceSelector>
  explicit queue(const context& syclContext,
                 const DeviceSelector& deviceSelector,
                 const async_handler& asyncHandler,
                 const property_list& propList = {});

  explicit queue(const context& syclContext, const device& syclDevice,
                 const property_list& propList = {});

  explicit queue(const context& syclContext, const device& syclDevice,
                 const async_handler& asyncHandler,
                 const property_list& propList = {});
```

Reply:
> From one point of view, we know how we going to use this code: only with one param and this check isn't really required. 

That's the view I take - it's an internal class which usage we fully control, and at this time we do not need more than one parameter. As of now, we do not need an abstract holder with an arbitrary number of parameters supported.

One specific thing with variadic pack ctor which I did not even realize is that it served as a default ctor for `__queue_holder` as well. Once I removed it, it became obvious that `device_policy() = default;` needs attention within this patch, and I have made it more solid.

Reply:
OK